modules/internal/ArrayViewSlice.chpl

```diff
@@ -0,0 +1,391 @@
+    //
+    // TODO: Can we privatize upon creation of the array-view slice and cache
+    // the results?
+    const _DomPid;
+    const dom; // Seems like the compiler requires a field called 'dom'...
+
+    // the representation of the sliced array
+    const _ArrPid;
+    const _ArrInstance;
+
+    // If this is an array view on a DefaultRectangular array
+    // (eventually...), the indexCache provides a mean of directly
+    // accessing the array's ddata to avoid indirection overheads
+    // through the array field above.
+    const indexCache = buildIndexCache();
+
+
+    //
+    // standard generic aspects of arrays
+    //
+
+    // these could be fields, but indirecting works just as well and
+    // makes the class less generic.
+    proc idxType type return dom.idxType;
+    proc rank param return arr.rank;
+
+    // The following seems like it ought to work, but it causes an
+    // error in the compiler for non-devel mode...  presumably due to
+    // a direct query of eltType in the compiler(?).  As a TODO we
+    // might want to hunt this down in the future...
+    //
+    //  proc eltType type return arr.eltType;
+
+
+    //
+    // introspection routine used elsewhere to filter array views
+    //
+    // TODO: Could this be replaced with more type-based introspection?
+    // I shied away from it since this is a generic class, but there
+    // must be (or should be) some way to do it without relying on
+    // methods like this...
+    //
+    proc isSliceArrayView() param {
+      return true;
+    }
+
+
+    //
+    // standard iterators
+    //
+
+    iter these() ref {
+      for i in privDom do
+        yield arr.dsiAccess(i);
+    }
+
+    iter these(param tag: iterKind) ref
+    where tag == iterKind.standalone && !localeModelHasSublocales {
+      for i in privDom.these(tag) do yield arr.dsiAccess(i);
+    }
+
+    iter these(param tag: iterKind) where tag == iterKind.leader {
+      for followThis in privDom.these(tag) do {
+        yield followThis;
+      }
+    }
+
+    iter these(param tag: iterKind, followThis) ref
+    where tag == iterKind.follower {
+      for i in privDom.these(tag, followThis) {
+        yield arr.dsiAccess[i];
+      }
+    }
+
+
+    //
+    // I/O
+    //
+
+    proc dsiSerialWrite(f) {
+      chpl_serialReadWriteRectangular(f, arr, privDom);
+    }
+
+    proc dsiSerialRead(f) {
+      chpl_serialReadWriteRectangular(f, arr, privDom);
+    }
+
+    proc dsiDisplayRepresentation() {
+      writeln("Slice view");
+      writeln("----------");
+      writeln("of domain:");
+      privDom.dsiDisplayRepresentation();
+      writeln("on array:");
+      arr.dsiDisplayRepresentation();
+      writeln("----------");
+    }
+
+
+    //
+    // accessors
+    //
+
+    inline proc dsiAccess(i: idxType ...rank) ref {
+      return dsiAccess(i);
+    }
+
+    inline proc dsiAccess(i: idxType ...rank)
+    where !shouldReturnRvalueByConstRef(eltType) {
```

Comment:
I like to indent where clauses but you might prefer this style. Just making sure it's not an accidental.